### ISSUE TITLE
High RAM Consumption when activating the WAF on multiple sites

### ISSUE THREAD

**Issue by skixmix** [corazawaf/coraza-caddy#76]
Hello, 

I'm encountering a problem with this module, specifically, the RAM usage keeps rising consistently whenever I incorporate a new Caddyfile configuration (or when I activate the WAF on an existing config).

My main configuration looks like this one:
```
{ Caddy global options... }
import *.caddy
```

In each Caddyfile, I manage a distinct site. I have approximately 200 Caddyfiles in total. Prior to installing the module, the loaded configuration utilized merely 1 GB of memory. However, once the module is installed and the WAF directive is enabled like this:

```
        coraza_waf {
			include /waf/coraza/coraza.conf-recommended
			include /waf/coreruleset/crs-setup.conf.example
			include /waf/coreruleset/rules/*.conf
			include /var/sites_waf/specific_site_configuration.conf
        }

```

I observe an increment of nearly 200 MB of RAM per site. Upon restarting the Caddy server, the RAM usage amplifies further and appears to remain unreleased. This does not occur when I deactivate the module and the WAF directive.

Do you have any idea on what is happening?

Thank you,
Simone

**Comment by jptosso** [comment]
Hey! Could you provide other metrics, like concurrent users and traffic information? Each WAF instance is just a few KBs of memory, it shouldn't create that much overhead.

It could be related to the garbage collector.

**Comment by skixmix** [comment]
Hello,

Thank you, as of now I'm testing it on a VM running Ubuntu 20.04.6 LTS and not exposed to the internet, so there are no clients and no traffic. On the other VMs in production, I have Caddy without the WAF module installed and the same configurations occupy more or less 1,3 GB of RAM. Whenever I perform a Caddy reload or restart, it takes up some additional memory, goes up to 2GB and then releases it back to 1,3 GB.

This behaviour changes in the test VM with the coraza module installed, and what happens is what I described before.

**Comment by jcchavezs** [comment]
I've been checking this issue but could not get my head around it. 

## Locally
When I build the binary `mage buildCaddy` and run it in local with the example caddyfile

```bash
# run httpbin in terminal 1
go run github.com/mccutchen/go-httpbin/v2/cmd/go-httpbin@v2.9.0 -port 8081
# run caddy in terminal 2, make sure you change the host in caddyfile from httpbin to localhost
./build/caddy run --config example/Caddyfile --adapter caddyfile
# run ab in terminal 3
ab -n 100000 -c 100 http://localhost:8080/
```

I get a caddy server going from 40MB to 75MB (checked with Activity Monitor)

## Using docker
If I run the same `ab` command but running the example in a container the memory can get up from 40MB to 4GB and slowly go down to 2GB (checked with docker stats). Both tests are similar, however the local one built the mac version and docker one uses the linux version (`GOOS=linux`). 

```bash
# run example
go run mage.go buildExample runExample
# run ab
ab -n 100000 -c 100 http://localhost:8080/
```

Any clue on why this could happen cc @anuraaga @mholt?

**Comment by mholt** [comment]
To know for sure, you'll need to capture a memory profile.

You can get one from `:2019/debug/pprof` on your server and viewing the allocations.

You can use `go tool pprof` to view them interactively, though sometimes it can be obvious just looking at the raw dump. I like generating an SVG to see what call stacks are allocating the most memory.

Here's an example tutorial (skip the top part that talks about adding the pprof handlers, Caddy already does this for you, hence the pprof endpoint you can load): https://www.freecodecamp.org/news/how-i-investigated-memory-leaks-in-go-using-pprof-on-a-large-codebase-4bec4325e192/

**Comment by jcchavezs** [comment]
Thanks @mholt, yeah I was doing profiling, just curious about the
difference between local and docker in case something came from the top of
your head.

**Comment by mholt** [comment]
Not really a fair comparison since the two are completely different platforms/OSes. Will need a profile to be sure.

**Comment by skixmix** [comment]
Hello,

Thank you for your support.  If it might be useful, what I'm observing is that each time I execute a `reboot` of the machine or a `systemctl restart` of the Caddy service, the RAM usage reverts to approximately 1 GB. After this operation, even if I execute multiple 'curl' requests on various sites, it remains stable and the usage does not increase by 1MB as it does in your situation. Additionally, as mentioned previously, I'm observing a rise in RAM utilization whenever I add a new site (or enable the WAF on a site where the directive was previously absent) and subsequently execute the rebuild/reload of Caddy configurations. 

Caddy version: `v2.6.4 h1:2hwYqiRwk1tf3VruhMpLcYTg+11fCdr8S3jhNAdnPy8=`
Coraza Caddy version: `1.2.2`

The VM is operating on vMware, not Docker, and on the same host there are two additional VMs without the Coraza module installed. All VMs share the same Caddy configurations (excluding the WAF directive, which is absent in the production VMs), and the same system: 

- OS: Ubuntu 20.04
- RAM: 4 GB (1.4-1.6 GB used in production, same on the test machine, but increasing like said before)
- CPU: 4 vCores (0.15 - 0.20 load average on production, 0.01 on the test machine)

**Comment by jcchavezs** [comment]
Could you try only these directives and check the memory?

```

        coraza_waf {
			include /waf/coraza/coraza.conf-recommended
			include /waf/coreruleset/crs-setup.conf.example
			include /waf/coreruleset/rules/*.conf
        }

```

**Comment by skixmix** [comment]
I tried replacing every site `coraza_waf` config with this one
```
        coraza_waf {
			include /waf/coraza/coraza.conf-recommended
			include /waf/coreruleset/crs-setup.conf.example
			include /waf/coreruleset/rules/*.conf
        }
```
without the site-specific configuration, and the problem persists even after reloading. The initial RAM usage was of 1.64 GB, which increased to 2.33 GB. After a few minutes, it released the memory, bringing it down to 1.88 GB. However, I'm unsure why it continues to retain that extra 200 MB of RAM.

**Comment by skixmix** [comment]
Hello, is there any update on this issue?

**Comment by jcchavezs** [comment]
Got the profiling

![profile001](https://github.com/corazawaf/coraza-caddy/assets/3075074/5ccceae2-6fd8-418b-98da-02d538feb1ae)

**Comment by jcchavezs** [comment]
I think one issue here is that we are initializing one instance of WAF per site, meaning that regexes and dicts (used by aho-corasick) are being initialized many times even when it is the same content and it is immutable. One idea to overcome this could be to use a regex map and dictionary map, however we need to be careful as tinygo won't support synchronization (hence non concurrent map for tinygo). This way, if you span 100 WAF with CRS, the regexes are going to use the same one in memory cc @anuraaga

**Comment by anuraaga** [comment]
I think a map for caching by pattern can work ok. Since TinyGo only runs on one thread, no worries about concurrency. One issue is all regex then are memory leaks - a pattern of replacing the waf instance with a different one with different rules becomes questionable. It's not common so maybe it doesn't need to be supported, but I guess making cache opt-in with a build tag is safer

**Comment by skixmix** [comment]
Hello,

If I got it right, there are two problems here:

1. When using the WAF module with multiple sites, it consumes a lot of memory because of regex allocation (duplicate rules saved in separate memory blocks for each site).
2. Every time we reload the Caddy configuration, the module appears to use up some RAM and then doesn't release all of it, even if nothing changed in the config.

Concerning the first problem, I can consider allocating additional RAM to the VM as a temporary solution. However, I'm unsure how to address the second issue. In the long run, it will take up all the RAM...
Do you have any updates or suggestions?

Thank you,
Simone

**Comment by jptosso** [comment]
Hello and thank you for the detailed analysis. Indeed we are working on a solution, as described by JC, we will implement cache mechanisms to avoid duplication on dictionaries for pm and regexes for rx. It will take some time to design this as it is a sensitive change. We will use this issue as reference so we will keep you posted.

Multiple hosts in a single web server is an important use case for coraza.

**Comment by jcchavezs** [comment]
@anuraaga how about we introduce a `Close` method on WAF where we delete all rules (and regexes) that can be used on reload. Is there any way to trigger a reload in a plugin @mholt ?

**Comment by anuraaga** [comment]
Having a Close is important, SGTM

**Comment by jcchavezs** [comment]
Do you consider it as a breaking change @anuraaga ?

**Comment by mholt** [comment]
I don't know the exact situation in the code, of course, but there's two simple rules I try to abide by:

1. Store all state in the module's struct itself; this will get garbage collected after a config reload. If it doesn't, then not "all state" is being stored in the module's struct.
2. If you do need global state, or state that persists across reloads, but is based on configuration that may or may not be needed after a reload, use a `UsagePool`: https://pkg.go.dev/github.com/caddyserver/caddy/v2#UsagePool -- often in conjunction with a module implementing the CleanerUpper: https://pkg.go.dev/github.com/caddyserver/caddy/v2#CleanerUpper -- this allows modules to say when they're done with values as configs are cycled through, and Caddy & Go's GC will take care of the rest.

**Comment by skixmix** [comment]
Hello everyone,

thanks for the prompt and diligent responses. Will the upcoming 2.0 module version include a resolution for this issue? If yes, do you have any timeline in mind?

Currently, I have Caddy deployed in production for various sites, with the WAF temporarily disabled. As a workaround, I've implemented a server-side (Node.js Express) WAF. However, I'm eager to integrate Coraza on Caddy as soon as possible.

Thank you,
Simone

**Comment by jcchavezs** [comment]
Version 2 has indeed to address this I would first try with singleton for
Regex before reusing a WAF because it only takes a single change in
directives to duplicate the memory used in regexes. Given that coraza-caddy
has no dedicated maintainer I would say the best timeline would be defined
by community (otherwise it goes to a queue of things that we maintainers
have to work on). This specific change is easy and we can get it under a
build tag before releasing it, we just need to add a singleton for when a
Regex is being compiled. Would you be up for such a change?

**Comment by jcchavezs** [comment]
I believe you can do:

```bash
XCADDY_GO_BUILD_FLAGS="-ldflags '-w s' -trimpath -tags my_custom_tag" xcaddy build --with github.com/corazawaf/coraza-caddy/v2
```

cc @mholt

**Comment by mholt** [comment]
Let me know how I can help :)

**Comment by jcchavezs** [comment]
> Store all state in the module's struct itself; this will get garbage collected after a config reload. If it doesn't, then not "all state" is being stored in the module's struct.

Is there any callback for when a module struct is going to be dismissed?

**Comment by mholt** [comment]
Good question: yes there is! :)

> a module implementing the CleanerUpper: https://pkg.go.dev/github.com/caddyserver/caddy/v2#CleanerUpper -- this allows modules to say when they're done with values as configs are cycled through, and Caddy & Go's GC will take care of the rest.

**Comment by jcchavezs** [comment]
Awesome! Thanks @mholt

**Comment by jcchavezs** [comment]
@skixmix I opened this PR https://github.com/corazawaf/coraza/pull/836. Please use that commit to give it a try and we see. You don't need to use a build tag.

**Comment by skixmix** [comment]
Great! Thank you so much. I'll give it a try as soon as possible and provide you with an update.

Have a nice day,
Simone

**Comment by skixmix** [comment]
Hello,

I have an update. What I did was to start with a clean Caddy setup, no site configured and no traffic. Then, I added 23 sites one by one, each one with its own Caddyfile and WAF configuration + OWASP CRS.
- RAM consumption before starting: 340MB
- RAM consumption after 10 sites: 1.20 GB
- RAM consumption after 23 sites: 2.06 GB

Throughout the process, I noticed that the RAM usage was having peaks at 3 GB and gradually dropping successively.
- to 1.8 GB (with 10 sites) and 
- to 2.8 GB (with 23 sites)

before reaching a stable consumption.

--- 

This is what happens when I add a new site or when I perform a reload of a site configuration (in this specific case, I added a new site).

### Initial RAM consumption
![image](https://github.com/corazawaf/coraza-caddy/assets/40468023/fa11ad52-e6d4-456e-b9b6-b3c5b9d35d5e)

### During configuration reload
![image](https://github.com/corazawaf/coraza-caddy/assets/40468023/a56a821c-e4c3-40a9-b6b1-1e47aab64ebd)

### After configuration reload
![image](https://github.com/corazawaf/coraza-caddy/assets/40468023/ef4d962b-8a70-480d-b004-3ee30593efe2)

### After more or less 3 minutes
![image](https://github.com/corazawaf/coraza-caddy/assets/40468023/97420b09-97b3-4cc0-9883-5e022ddcab94)

### After 30 minutes
![image](https://github.com/corazawaf/coraza-caddy/assets/40468023/784187f2-5c14-482a-be59-44063dd0775d)

---

## Production comparison

This is the current state of one of the VMs that I have running on one of my DCs in production, with more than 100 site configurations, same as the test ones, but without the WAF module:
![image](https://github.com/corazawaf/coraza-caddy/assets/40468023/b1906897-d979-4877-b84a-e5f706ed7d30)

--

# Recap
The problem persists in such a manner that each configuration uses over 100 MB of RAM instead of the 8/9 MB used in production without the WAF, which is 10 times higher. Nevertheless, I believe that one of the issues has been resolved as the memory usage now decreases rather than continuously leaking indefinitely. 😄

**Comment by mholt** [comment]
You can find out what's using memory by capturing a profile. Caddy exposes HTTP endpoints for this at :2019/debug/pprof -- this is a simple HTML document with links to collect various profiles, either instantaneously or over a time period if you add query params. Capture the heap profile then use `go tool pprof ...` to open the profile; type `top` to see the top allocations, or `svg` to generate a vector image you can view in your browser.

**Comment by jcchavezs** [comment]
According to https://github.com/corazawaf/coraza-caddy/issues/76#issuecomment-1602251630 we might need to memoize the aho-corasick dictionaries too.

**Comment by anamba** [comment]
I believe I am running into this issue... on startup, Caddy consumes about 1.5g ram, but each successive `caddy reload` (when there are config changes to apply) appears to add on about another 1.5g. Since the machine in question has only 4g physical ram, this becomes a problem quickly.

The memory usage does appear to come back down somewhat about 3-5 minutes later, but unlike the graphs, above I am not seeing memory usage level off, it just seems to settle at a higher level each time (eventually getting either OOMed, or restarted by monit when :2019 becomes slow to respond).

The high memory consumption in general is also a problem... this server has 90 sites, and I would like to enable coraza on all of them eventually, but that is currently not possible with only 4G ram. (It is a dev server, I am trying to work out the optimal coraza config here before enabling coraza on the production equivalents.)

The behavior I have described here is based on a build using the PR branch mentioned above in https://github.com/corazawaf/coraza-caddy/issues/76#issuecomment-1622509997.

**Comment by skixmix** [comment]
Hi, do you have any update about this issue? Also, for v2, there is this one still pending #88 

Thanks you,
Simone

**Comment by jptosso** [comment]
Hey, we have a new feature that should improve this, if you can build your own Caddy that would help testing:
```sh
XCADDY_GO_BUILD_FLAGS='-tags=memoize_builders' xcaddy build --with github.com/corazawaf/coraza-caddy/v2
```

If you are using Docker you can build it using
```dockerfile
FROM caddy:2.7-builder
XCADDY_GO_BUILD_FLAGS='-tags=memoize_builders' xcaddy build --with github.com/corazawaf/coraza-caddy/v2
```

**Comment by anamba** [comment]
More testing needed, but initial tests show a huge decrease in memory usage. Promising!

**Comment by skixmix** [comment]
Hello,

I've tested the new feature with `memoize_builders` and performed the same experiment of https://github.com/corazawaf/coraza-caddy/issues/76#issuecomment-1623471577. It now seems to go from 340 MB of RAM consumption to 540 MB with 10 sites, so it looks really promising!

Also, with this version #88 is no more an issue.

Thank you for the awesome work!
Simone

**Comment by jptosso** [comment]
Great news! @jcchavezs do you think we should enable this feature by default?

**Comment by jcchavezs** [comment]
@skixmix more memory improvements are coming. I am working on another PR to lazily load regexes.

@jptosso maybe we can enable it but we need to make sure risks are well documented, although minimal.

**Comment by fzipi** [comment]
We are pushing to enable memoize as the new default now, with the option to _disable_ in a build flag. See https://github.com/corazawaf/coraza/pull/1540.

**Comment by Forza-tng** [comment]
Hi,

I just had the same issue when enabling Coraza WAF on a Caddy server on multiple hosts. Memory usage grew from about 300MiB to over 30GiB and kept increasing, almost killing the server as all ram and all swap got eaten up.

I use this config:

```
	coraza_waf {
		load_owasp_crs
			directives `
			Include @coraza.conf-recommended
			Include @crs-setup.conf.example
			
			Include /etc/caddy/crs/plugins/*-config.conf
			Include /etc/caddy/crs/plugins/*-before.conf
			
			Include @owasp_crs/*.conf
			
			Include /etc/caddy/crs/plugins/*-after.conf
			
			SecAuditEngine RelevantOnly
			SecAuditLog /var/log/caddy/coraza_audit_REDACTED.log
			SecAuditLogParts ABCFHKZ
			SecRuleEngine Off
		`
	}
```

I'm using caddy 2.11.3 and built it today using xcaddy.
```shell
extras="--with github.com/caddyserver/caddy/v2=/usr/src/caddy/git/caddy \
--with github.com/caddyserver/transform-encoder \
--with github.com/dunglas/caddy-cbrotli \
--with github.com/WeidiDeng/caddy-cloudflare-ip \
--with github.com/caddyserver/forwardproxy \
--with github.com/fvbommel/caddy-combine-ip-ranges \
--with github.com/corazawaf/coraza-caddy/v2 "

xcaddy build ${extras}
```

I have the nextcloud plugin in `/etc/caddy/crs/plugins`.
https://github.com/coreruleset/nextcloud-rule-exclusions-plugin

**Comment by jptosso** [comment]
Can you check you are using the latest version of coraza caddy? 
Also can you share a profile of the memory ? 

https://caddyserver.com/docs/profiling
It doesn't contain sensitive data, just memory allocation in numbers

**Comment by Forza-tng** [comment]
The build used coraza caddy v2.5.0:
<details>
<summary>❯ caddy build-info</summary>

```
go      go1.26.2-X:nodwarf5
path    caddy
mod     caddy   (devel)
dep     cel.dev/expr    v0.25.1 h1:1KrZg61W6TWSxuNZ37Xy49ps13NUovb66QLprthtwi4=
dep     cloud.google.com/go/auth        v0.18.2 h1:+Nbt5Ev0xEqxlNjd6c+yYUeosQ5TtEUaNcN/3FozlaM=
dep     cloud.google.com/go/auth/oauth2adapt    v0.2.8  h1:keo8NaayQZ6wimpNSmW5OPc283g65QNIiLpZnkHRbnc=
dep     cloud.google.com/go/compute/metadata    v0.9.0  h1:pDUj4QMoPejqq20dK0Pg2N4yG9zIkYGdBtwLoEkH9Zs=
dep     dario.cat/mergo v1.0.2  h1:85+piFYR1tMbRrLcDwR18y4UKJ3aH1Tbzi24VRW1TK8=
dep     filippo.io/bigmod       v0.1.0  h1:UNzDk7y9ADKST+axd9skUpBQeW7fG2KrTZyOE4uGQy8=
dep     filippo.io/edwards25519 v1.2.0  h1:crnVqOiS4jqYleHd9vaKZ+HKtHfllngJIiOpNpoJsjo=
dep     github.com/BurntSushi/toml      v1.6.0  h1:dRaEfpa2VI55EwlIW72hMRHdWouJeRF7TPYhI+AUQjk=
dep     github.com/DeRuina/timberjack   v1.4.2  h1:4bKlzhKdsR+2oNkgef9mqb4n11ICow8VK88RfzJPzN8=
dep     github.com/KimMachineGun/automemlimit   v0.7.5  h1:RkbaC0MwhjL1ZuBKunGDjE/ggwAX43DwZrJqVwyveTk=
dep     github.com/Masterminds/goutils  v1.1.1  h1:5nUrii3FMTL5diU80unEVvNevw1nH4+ZV4DSLVJLSYI=
dep     github.com/Masterminds/semver/v3        v3.4.0  h1:Zog+i5UMtVoCU8oKka5P7i9q9HgrJeGzI9SA1Xbatp0=
dep     github.com/Masterminds/sprig/v3 v3.3.0  h1:mQh0Yrg1XPo6vjYXgtf5OtijNAKJRNcTdOOGZe3tPhs=
dep     github.com/WeidiDeng/caddy-cloudflare-ip        v0.0.0-20231130002422-f53b62aa13cb      h1:0lhny8EXcEeOWwGeZFkWrnPeSUEyqGwLH4Q0hCIKrrg=
dep     github.com/alecthomas/chroma/v2 v2.23.1 h1:nv2AVZdTyClGbVQkIzlDm/rnhk1E9bU9nXwmZ/Vk/iY=
dep     github.com/antlr4-go/antlr/v4   v4.13.1 h1:SqQKkuVZ+zWkMMNkjy5FZe5mr5WURWnlpmOuzYWrPrQ=
dep     github.com/aryann/difflib       v0.0.0-20210328193216-ff5ff6dc229b      h1:uUXgbcPDK3KpW29o4iy7GtuappbWT0l5NaMo9H9pJDw=
dep     github.com/beorn7/perks v1.0.1  h1:VlbKKnNfV8bJzeqoa4cOKqO6bYr3WgKZxO8Z16+hsOM=
dep     github.com/buger/jsonparser     v1.1.2  h1:frqHqw7otoVbk5M8LlE/L7HTnIq2v9RX6EJ48i9AxJk=
dep     github.com/caddyserver/caddy/v2 v2.11.3
=>      /usr/src/caddy/git/caddy        (devel)
dep     github.com/caddyserver/certmagic        v0.25.3 h1:mGf5ba8F7xA4c5jfDZZbK2buY1VEkbnwpMDixaju94A=
dep     github.com/caddyserver/forwardproxy     v0.0.0-20260321230143-0aab84dad4fc      h1:hr5UPoMON9LBlnrS/tCllhSSn5EK7zkOS2BN9u0If/0=
dep     github.com/caddyserver/transform-encoder        v0.0.0-20260423033309-ba4124974830      h1:jskaH5w734LNiRXVrbtPd9vGq9G2/1vV0JDgM/VsiXQ=
dep     github.com/caddyserver/zerossl  v0.1.5  h1:dkvOjBAEEtY6LIGAHei7sw2UgqSD6TrWweXpV7lvEvE=
dep     github.com/ccoveille/go-safecast/v2     v2.0.0  h1:+5eyITXAUj3wMjad6cRVJKGnC7vDS55zk0INzJagub0=
dep     github.com/cenkalti/backoff/v5  v5.0.3  h1:ZN+IMa753KfX5hd8vVaMixjnqRZ3y8CuJKRKj1xcsSM=
dep     github.com/cespare/xxhash/v2    v2.3.0  h1:UL815xU9SqsFlibzuggzjXhog7bL6oX9BbNZnL2UFvs=
dep     github.com/chzyer/readline      v1.5.1  h1:upd/6fQk4src78LMRzh5vItIt361/o4uq553V8B5sGI=
dep     github.com/cloudflare/circl     v1.6.3  h1:9GPOhQGF9MCYUeXyMYlqTR6a5gTrgR/fBLXvUgtVcg8=
dep     github.com/corazawaf/coraza-caddy/v2    v2.5.0  h1:zdK/o14duUII/UCeJo5yozN1MhinASGULrJgQT+GIG4=
dep     github.com/corazawaf/coraza-coreruleset/v4      v4.25.0 h1:tqFO1lfVpTiyWtlN618OXpZMfw+nnN0Q4///W5W+/HM=
dep     github.com/corazawaf/coraza/v3  v3.7.0  h1:LIQqu1r+l6e/U/gyiZeykWaNNBY1TzRLz+aaI+QYEEM=
dep     github.com/corazawaf/libinjection-go    v0.3.2  h1:9rrKt0lpg4WvUXt+lwS06GywfqRXXsa/7JcOw5cQLwI=
dep     github.com/coreos/go-oidc/v3    v3.17.0 h1:hWBGaQfbi0iVviX4ibC7bk8OKT5qNr4klBaCHVNvehc=
dep     github.com/cpuguy83/go-md2man/v2        v2.0.7  h1:zbFlGlXEAKlwXpmvle3d8Oe3YnkKIK4xSRTd3sHPnBo=
dep     github.com/dlclark/regexp2      v1.11.5 h1:Q/sSnsKerHeCkc/jSTNq1oCm7KiVgUMZRDUoRu0JQZQ=
dep     github.com/dunglas/caddy-cbrotli        v1.0.1  h1:mkg7EB1GmoyfBt3kY3mq4o/0bfnBeq7ZLQjmVmdBE3Y=
dep     github.com/dustin/go-humanize   v1.0.1  h1:GzkhY7T5VNhEkwH0PVJgjz+fX1rhBrR7pRT3mDkpeCY=
dep     github.com/felixge/httpsnoop    v1.0.4  h1:NFTV2Zj1bL4mc9sqWACXbQFVBBg2W3GPvqp8/ESS2Wg=
dep     github.com/fvbommel/caddy-combine-ip-ranges     v0.0.1  h1:xRknvJHfip/zG5XarsGoG5wvssdA1lMftjGQBAc2W0o=
dep     github.com/fxamacker/cbor/v2    v2.9.0  h1:NpKPmjDBgUfBms6tr6JZkTHtfFGcMKsw3eGcmD/sapM=
dep     github.com/go-chi/chi/v5        v5.2.5  h1:Eg4myHZBjyvJmAFjFvWgrqDTXFyOzjj7YIm3L3mu6Ug=
dep     github.com/go-jose/go-jose/v3   v3.0.4  h1:Wp5HA7bLQcKnf6YYao/4kpRpVMp/yf6+pJKV8WFSaNY=
dep     github.com/go-jose/go-jose/v4   v4.1.4  h1:moDMcTHmvE6Groj34emNPLs/qtYXRVcd6S7NHbHz3kA=
dep     github.com/go-logr/logr v1.4.3  h1:CjnDlHq8ikf6E492q6eKboGOC0T8CDaOvkHCIg8idEI=
dep     github.com/go-logr/stdr v1.2.2  h1:hSWxHoqTgW2S2qGc0LTAI563KZ5YKYRhT3MFKZMbjag=
dep     github.com/goccy/go-json        v0.10.5 h1:Fq85nIqj+gXn/S5ahsiTlK3TmC85qgirsdTP/+DeaC4=
dep     github.com/goccy/go-yaml        v1.18.0 h1:8W7wMFS12Pcas7KU+VVkaiCng+kG8QiFeFwzFb+rwuw=
dep     github.com/google/brotli/go/cbrotli     v1.1.0  h1:YwHD/rwSgUSL4b2S3ZM2jnNymm+tmwKQqjUIC63nmHU=
dep     github.com/google/cel-go        v0.28.0 h1:KjSWstCpz/MN5t4a8gnGJNIYUsJRpdi/r97xWDphIQc=
dep     github.com/google/certificate-transparency-go   v1.1.8-0.20240110162603-74a5dd331745    h1:heyoXNxkRT155x4jTAiSv5BVSVkueifPUm+Q8LUXMRo=
dep     github.com/google/go-tpm        v0.9.8  h1:slArAR9Ft+1ybZu0lBwpSmpwhRXaa85hWtMinMyRAWo=
dep     github.com/google/go-tspi       v0.3.0  h1:ADtq8RKfP+jrTyIWIZDIYcKOMecRqNJFOew2IT0Inus=
dep     github.com/google/s2a-go        v0.1.9  h1:LGD7gtMgezd8a/Xak7mEWL0PjoTQFvpRudN895yqKW0=
dep     github.com/google/uuid  v1.6.0  h1:NIvaJDMOsjHA8n1jAhLSgzrAzy1Hgr+hNrb57e+94F0=
dep     github.com/googleapis/enterprise-certificate-proxy      v0.3.14 h1:yh8ncqsbUY4shRD5dA6RlzjJaT4hi3kII+zYw8wmLb8=
dep     github.com/googleapis/gax-go/v2 v2.18.0 h1:jxP5Uuo3bxm3M6gGtV94P4lliVetoCB4Wk2x8QA86LI=
dep     github.com/gotnospirit/makeplural       v0.0.0-20180622080156-a5f48d94d976      h1:b70jEaX2iaJSPZULSUxKtm73LBfsCrMsIlYCUgNGSIs=
dep     github.com/gotnospirit/messageformat    v0.0.0-20221001023931-dfe49f1eb092      h1:c7gcNWTSr1gtLp6PyYi3wzvFCEcHJ4YRobDgqmIgf7Q=
dep     github.com/grpc-ecosystem/grpc-gateway/v2       v2.28.0 h1:HWRh5R2+9EifMyIHV7ZV+MIZqgz+PMpZ14Jynv3O2Zs=
dep     github.com/huandu/xstrings      v1.5.0  h1:2ag3IFq9ZDANvthTwTiqSSZLjDc+BedvHPAp5tJy2TI=
dep     github.com/jcchavezs/mergefs    v0.1.1  h1:D45R17m6dHnSVZefnhynoeZvcK2Uw0oTrRfoUOQ0S5Y=
dep     github.com/kaptinlin/go-i18n    v0.1.4  h1:wCiwAn1LOcvymvWIVAM4m5dUAMiHunTdEubLDk4hTGs=
dep     github.com/kaptinlin/jsonschema v0.4.6  h1:vOSFg5tjmfkOdKg+D6Oo4fVOM/pActWu/ntkPsI1T64=
dep     github.com/klauspost/compress   v1.18.5 h1:/h1gH5Ce+VWNLSWqPzOVn6XBO+vJbCNGvjoaGBFW2IE=
dep     github.com/klauspost/cpuid/v2   v2.3.0  h1:S4CRMLnYUhGeDFDqkGriYKdfoFlDnMtqTiI/sFzhA9Y=
dep     github.com/libdns/libdns        v1.1.1  h1:wPrHrXILoSHKWJKGd0EiAVmiJbFShguILTg9leS/P/U=
dep     github.com/manifoldco/promptui  v0.9.0  h1:3V4HzJk1TtXW1MTZMP7mdlwbBpIinw3HztaIlYthEiA=
dep     github.com/mattn/go-colorable   v0.1.14 h1:9A9LHSqF/7dyVVX6g0U9cwm9pG3kP9gSzcuIPHPsaIE=
dep     github.com/mattn/go-isatty      v0.0.20 h1:xfD0iDuEKnDkl03q4limB+vH+GxLEtL/jb4xVJSWWEY=
dep     github.com/mgutz/ansi   v0.0.0-20200706080929-d51e80ef957d      h1:5PJl274Y63IEHC+7izoQE9x6ikvDFZS2mDVS3drnohI=
dep     github.com/mholt/acmez/v3       v3.1.6  h1:eGVQNObP0pBN4sxqrXeg7MYqTOWyoiYpQqITVWlrevk=
dep     github.com/miekg/dns    v1.1.72 h1:vhmr+TF2A3tuoGNkLDFK9zi36F2LS+hKTRW0Uf8kbzI=
dep     github.com/mitchellh/copystructure      v1.2.0  h1:vpKXTN4ewci03Vljg/q9QvCGUDttBOGBIa15WveJJGw=
dep     github.com/mitchellh/go-ps      v1.0.0  h1:i6ampVEEF4wQFF+bkYfwYgY+F/uYJDktmvLPf7qIgjc=
dep     github.com/mitchellh/reflectwalk        v1.0.2  h1:G2LzWKi524PWgd3mLHV8Y5k7s6XUvT0Gef6zxSIeXaQ=
dep     github.com/munnerz/goautoneg    v0.0.0-20191010083416-a7dc8b61c822      h1:C3w9PqII01/Oq1c1nUAm88MOHcQC9l5mIlSMApZMrHA=
dep     github.com/pbnjay/memory        v0.0.0-20210728143218-7b4eea64cf58      h1:onHthvaw9LFnH4t2DcNVpwGmV9E1BkGknEliJkfwQj0=
dep     github.com/petar-dambovaliev/aho-corasick       v0.0.0-20250424160509-463d218d4745      h1:Vpr4VgAizEgEZsaMohpw6JYDP+i9Of9dmdY4ufNP6HI=
dep     github.com/pires/go-proxyproto  v0.11.0 h1:gUQpS85X/VJMdUsYyEgyn59uLJvGqPhJV5YvG68wXH4=
dep     github.com/pkg/errors   v0.9.1  h1:FEBLx1zS214owpjy7qsBeixbURkuhQAwrK5UwLGTwt4=
dep     github.com/prometheus/client_golang     v1.23.2 h1:Je96obch5RDVy3FDMndoUsjAhG5Edi49h0RJWRi/o0o=
dep     github.com/prometheus/client_model      v0.6.2  h1:oBsgwpGs7iVziMvrGhE53c/GrLUsZdHnqNwqPLxwZyk=
dep     github.com/prometheus/common    v0.67.5 h1:pIgK94WWlQt1WLwAC5j2ynLaBRDiinoAb86HZHTUGI4=
dep     github.com/prometheus/otlptranslator    v1.0.0  h1:s0LJW/iN9dkIH+EnhiD3BlkkP5QVIUVEoIwkU+A6qos=
dep     github.com/prometheus/procfs    v0.20.1 h1:XwbrGOIplXW/AU3YhIhLODXMJYyC1isLFfYCsTEycfc=
dep     github.com/quic-go/qpack        v0.6.0  h1:g7W+BMYynC1LbYLSqRt8PBg5Tgwxn214ZZR34VIOjz8=
dep     github.com/quic-go/quic-go      v0.59.1 h1:0Gmua0HW1Tv7ANR7hUYwRyD0MG5OJfgvYSZasGZzBic=
dep     github.com/rs/xid       v1.6.0  h1:fV591PaemRlL6JfRxGDEPl69wICngIQ3shQtzfy2gxU=
dep     github.com/russross/blackfriday/v2      v2.1.0  h1:JIOH55/0cWyOuilr9/qlrm0BSXldqnqwMsf35Ld67mk=
dep     github.com/shopspring/decimal   v1.4.0  h1:bxl37RwXBklmTi0C79JfXCEBD1cqqHt0bbgBAGFp81k=
dep     github.com/shurcooL/sanitized_anchor_name       v1.0.0  h1:PdmoCO6wvbs+7yrJyMORt4/BmY5IYyJwS/kOiWx8mHo=
dep     github.com/sirupsen/logrus      v1.9.4  h1:TsZE7l11zFCLZnZ+teH4Umoq5BhEIfIzfRDZ1Uzql2w=
dep     github.com/slackhq/nebula       v1.10.3 h1:EstYj8ODEcv6T0R9X5BVq1zgWZnyU5gtPzk99QF1PMU=
dep     github.com/smallstep/certificates       v0.30.2 h1:1G3xBi8sJ740iA1mMPW2Svv7EIZKJ4Zf/iQtA5QlN0Y=
dep     github.com/smallstep/cli-utils  v0.12.2 h1:lGzM9PJrH/qawbzMC/s2SvgLdJPKDWKwKzx9doCVO+k=
dep     github.com/smallstep/go-attestation     v0.4.4-0.20241119153605-2306d5b464ca    h1:VX8L0r8vybH0bPeaIxh4NQzafKQiqvlOn8pmOXbFLO4=
dep     github.com/smallstep/linkedca   v0.25.0 h1:txT9QHGbCsJq0MhAghBq7qhurGY727tQuqUi+n4BVBo=
dep     github.com/smallstep/nosql      v0.8.0  h1:FBTCUfKPmWYbrozW+RBKu+fnvbn+zr5rVli/XB4Jp4A=
dep     github.com/smallstep/pkcs7      v0.2.1  h1:6Kfzr/QizdIuB6LSv8y1LJdZ3aPSfTNhTLqAx9CTLfA=
dep     github.com/smallstep/scep       v0.0.0-20250318231241-a25cabb69492      h1:k23+s51sgYix4Zgbvpmy+1ZgXLjr4ZTkBTqXmpnImwA=
dep     github.com/smallstep/truststore v0.13.0 h1:90if9htAOblavbMeWlqNLnO9bsjjgVv2hQeQJCi/py4=
dep     github.com/spf13/cast   v1.7.0  h1:ntdiHjuueXFgm5nzDRdOS4yfT43P5Fnud6DH50rz/7w=
dep     github.com/spf13/cobra  v1.10.2 h1:DMTTonx5m65Ic0GOoRY2c16WCbHxOOw6xxezuLaBpcU=
dep     github.com/spf13/pflag  v1.0.10 h1:4EBh2KAYBwaONj6b2Ye1GiHfwjqyROoF4RwYO+vPwFk=
dep     github.com/tailscale/tscert     v0.0.0-20251216020129-aea342f6d747      h1:RnBbFMmodYzhC6adOjTbtUQXyzV8dcvKYbolzs6Qch0=
dep     github.com/tidwall/gjson        v1.18.0 h1:FIDeeyB800efLX89e5a8Y0BNH+LOngJyGrIWxG2FKQY=
dep     github.com/tidwall/match        v1.1.1  h1:+Ho715JplO36QYgwN9PGYNhgZvoUSc9X2c80KVTi+GA=
dep     github.com/tidwall/pretty       v1.2.1  h1:qjsOFOWWQl+N3RsoF5/ssm1pHmJJwhjlSbZ51I6wMl4=
dep     github.com/urfave/cli   v1.22.17        h1:SYzXoiPfQjHBbkYxbew5prZHS1TOLT3ierW8SYLqtVQ=
dep     github.com/valllabh/ocsf-schema-golang  v1.0.3  h1:eR8k/3jP/OOqB8LRCtdJ4U+vlgd/gk5y3KMXoodrsrw=
dep     github.com/x448/float16 v0.8.4  h1:qLwI1I70+NjRFUR3zs1JPUCgaCXSh3SW62uAKT1mSBM=
dep     github.com/yuin/goldmark        v1.8.2  h1:kEGpgqJXdgbkhcOgBxkC0X0PmoPG1ZyoZ117rDVp4zE=
dep     github.com/yuin/goldmark-highlighting/v2        v2.0.0-20230729083705-37449abec8cc      h1:+IAOyRda+RLrxa1WC7umKOZRsGq4QrFFMYApOeHzQwQ=
dep     github.com/zeebo/blake3 v0.2.4  h1:KYQPkhpRtcqh0ssGYcKLG1JYvddkEA8QwCM/yBqhaZI=
dep     go.etcd.io/bbolt        v1.4.3  h1:dEadXpI6G79deX5prL3QRNP6JB8UxVkqo4UPnHaNXJo=
dep     go.opentelemetry.io/auto/sdk    v1.2.1  h1:jXsnJ4Lmnqd11kwkBV2LgLoFMZKizbCi5fNZ/ipaZ64=
dep     go.opentelemetry.io/contrib/bridges/prometheus  v0.68.0 h1:w3zlHYETbDwXyWHZlyyR58ZC39XGi8rAhkBgUgJ9d5w=
dep     go.opentelemetry.io/contrib/exporters/autoexport        v0.65.0 h1:2gApdml7SznX9szEKFjKjM4qGcGSvAybYLBY319XG3g=
dep     go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp   v0.65.0 h1:7iP2uCb7sGddAr30RRS6xjKy7AZ2JtTOPA3oolgVSw8=
dep     go.opentelemetry.io/contrib/propagators/autoprop        v0.65.0 h1:kTaCycF9Xkm8VBBvH0rJ4wFeRjtIV55Erk3uuVsIs5s=
dep     go.opentelemetry.io/contrib/propagators/aws     v1.43.0 h1:EwnsB3cXRLAh7/Nr/9rMuGw73nfb3z6uAvVDjRrbeUg=
dep     go.opentelemetry.io/contrib/propagators/b3      v1.43.0 h1:CETqV3QLLPTy5yNrqyMr41VnAOOD4lsRved7n4QG00A=
dep     go.opentelemetry.io/contrib/propagators/jaeger  v1.43.0 h1:peiLMz1+aqJE+3L4mOVtR9wlmv+yh/JVYXCBjqmzJJE=
dep     go.opentelemetry.io/contrib/propagators/ot      v1.43.0 h1:Hh1HahlGc81AOE7siqi1tVOlbanY/UxMMWedpb0d5oQ=
dep     go.opentelemetry.io/otel        v1.43.0 h1:mYIM03dnh5zfN7HautFE4ieIig9amkNANT+xcVxAj9I=
dep     go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploggrpc     v0.19.0 h1:Dn8rkudDzY6KV9dr/D/bTUuWgqDf9xe0rr4G2elrn0Y=
dep     go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp     v0.19.0 h1:HIBTQ3VO5aupLKjC90JgMqpezVXwFuq6Ryjn0/izoag=
dep     go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc       v1.43.0 h1:8UQVDcZxOJLtX6gxtDt3vY2WTgvZqMQRzjsqiIHQdkc=
dep     go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp       v1.43.0 h1:w1K+pCJoPpQifuVpsKamUdn9U0zM3xUziVOqsGksUrY=
dep     go.opentelemetry.io/otel/exporters/otlp/otlptrace       v1.43.0 h1:88Y4s2C8oTui1LGM6bTWkw0ICGcOLCAI5l6zsD1j20k=
dep     go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc v1.43.0 h1:RAE+JPfvEmvy+0LzyUA25/SGawPwIUbZ6u0Wug54sLc=
dep     go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp v1.43.0 h1:3iZJKlCZufyRzPzlQhUIWVmfltrXuGyfjREgGP3UUjc=
dep     go.opentelemetry.io/otel/exporters/prometheus   v0.65.0 h1:jOveH/b4lU9HT7y+Gfamf18BqlOuz2PWEvs8yM7Q6XE=
dep     go.opentelemetry.io/otel/exporters/stdout/stdoutlog     v0.19.0 h1:GJkybS+crDMdExT/BUNCEgfrmfboztcS6PhvSo88HKM=
dep     go.opentelemetry.io/otel/exporters/stdout/stdoutmetric  v1.43.0 h1:TC+BewnDpeiAmcscXbGMfxkO+mwYUwE/VySwvw88PfA=
dep     go.opentelemetry.io/otel/exporters/stdout/stdouttrace   v1.43.0 h1:mS47AX77OtFfKG4vtp+84kuGSFZHTyxtXIN269vChY0=
dep     go.opentelemetry.io/otel/log    v0.19.0 h1:KUZs/GOsw79TBBMfDWsXS+KZ4g2Ckzksd1ymzsIEbo4=
dep     go.opentelemetry.io/otel/metric v1.43.0 h1:d7638QeInOnuwOONPp4JAOGfbCEpYb+K6DVWvdxGzgM=
dep     go.opentelemetry.io/otel/sdk    v1.43.0 h1:pi5mE86i5rTeLXqoF/hhiBtUNcrAGHLKQdhg4h4V9Dg=
dep     go.opentelemetry.io/otel/sdk/log        v0.19.0 h1:scYVLqT22D2gqXItnWiocLUKGH9yvkkeql5dBDiXyko=
dep     go.opentelemetry.io/otel/sdk/metric     v1.43.0 h1:S88dyqXjJkuBNLeMcVPRFXpRw2fuwdvfCGLEo89fDkw=
dep     go.opentelemetry.io/otel/trace  v1.43.0 h1:BkNrHpup+4k4w+ZZ86CZoHHEkohws8AY+WTX09nk+3A=
dep     go.opentelemetry.io/proto/otlp  v1.10.0 h1:IQRWgT5srOCYfiWnpqUYz9CVmbO8bFmKcwYxpuCSL2g=
dep     go.step.sm/crypto       v0.77.1 h1:4EEqfKdv0egQ1lqz2RhnU8Jv6QgXZfrgoxWMqJF9aDs=
dep     go.uber.org/automaxprocs        v1.6.0  h1:O3y2/QNTOdbF+e/dpXNNW7Rx2hZ4sTIPyybbxyNqTUs=
dep     go.uber.org/multierr    v1.11.0 h1:blXXJkSxSSfBVBlC76pxqeO+LN3aDfLQo+309xJstO0=
dep     go.uber.org/zap v1.27.1 h1:08RqriUEv8+ArZRYSTXy1LeBScaMpVSTBhCeaZYfMYc=
dep     go.uber.org/zap/exp     v0.3.0  h1:6JYzdifzYkGmTdRR59oYH+Ng7k49H9qVpWwNSsGJj3U=
dep     go.yaml.in/yaml/v2      v2.4.4  h1:tuyd0P+2Ont/d6e2rl3be67goVK4R6deVxCUX5vyPaQ=
dep     go.yaml.in/yaml/v3      v3.0.4  h1:tfq32ie2Jv2UxXFdLJdh3jXuOzWiL1fo0bu/FbuKpbc=
dep     golang.org/x/crypto     v0.50.0 h1:zO47/JPrL6vsNkINmLoo/PH1gcxpls50DNogFvB5ZGI=
dep     golang.org/x/crypto/x509roots/fallback  v0.0.0-20260213171211-a408498e5541      h1:FmKxj9ocLKn45jiR2jQMwCVhDvaK7fKQFzfuT9GvyK8=
dep     golang.org/x/exp        v0.0.0-20251023183803-a4bb9ffd2546      h1:mgKeJMpvi0yx/sU5GsxQ7p6s2wtOnGAHZWCHUM4KGzY=
dep     golang.org/x/net        v0.53.0 h1:d+qAbo5L0orcWAr0a9JweQpjXF19LMXJE8Ey7hwOdUA=
dep     golang.org/x/oauth2     v0.36.0 h1:peZ/1z27fi9hUOFCAZaHyrpWG5lwe0RJEEEeH0ThlIs=
dep     golang.org/x/sync       v0.20.0 h1:e0PTpb7pjO8GAtTs2dQ6jYa5BWYlMuX047Dco/pItO4=
dep     golang.org/x/sys        v0.43.0 h1:Rlag2XtaFTxp19wS8MXlJwTvoh8ArU6ezoyFsMyCTNI=
dep     golang.org/x/term       v0.42.0 h1:UiKe+zDFmJobeJ5ggPwOshJIVt6/Ft0rcfrXZDLWAWY=
dep     golang.org/x/text       v0.36.0 h1:JfKh3XmcRPqZPKevfXVpI1wXPTqbkE5f7JA92a55Yxg=
dep     golang.org/x/time       v0.15.0 h1:bbrp8t3bGUeFOx08pvsMYRTCVSMk89u4tKbNOZbp88U=
dep     google.golang.org/api   v0.271.0        h1:cIPN4qcUc61jlh7oXu6pwOQqbJW2GqYh5PS6rB2C/JY=
dep     google.golang.org/genproto/googleapis/api       v0.0.0-20260401024825-9d38bb4040a9      h1:VPWxll4HlMw1Vs/qXtN7BvhZqsS9cdAittCNvVENElA=
dep     google.golang.org/genproto/googleapis/rpc       v0.0.0-20260401024825-9d38bb4040a9      h1:m8qni9SQFH0tJc1X0vmnpw/0t+AImlSvp30sEupozUg=
dep     google.golang.org/grpc  v1.80.0 h1:Xr6m2WmWZLETvUNvIUmeD5OAagMw3FiKmMlTdViWsHM=
dep     google.golang.org/protobuf      v1.36.11        h1:fV6ZwhNocDyBLK0dj+fg8ektcVegBBuEolpbTQyBNVE=
dep     gopkg.in/yaml.v3        v3.0.1  h1:fxVm/GzAzEWqLHuvctI91KS9hhNmmWOoWu0XTYJS7CA=
dep     rsc.io/binaryregexp     v0.2.0  h1:HfqmD5MEmC0zvwBuF187nq9mdnXjXsSivRiXN7SmRkE=
build   -buildmode=exe
build   -compiler=gc
build   -tags=nobadger,nomysql,nopgx
build   -trimpath=true
build   CGO_ENABLED=1
build   GOARCH=amd64
build   GOEXPERIMENT=nodwarf5
build   GOOS=linux
build   GOAMD64=v3
```

</details>

It's a little tricky to trigger the memory explosion, and when it happens it only takes a short while before it starts to eat up all ram (and gets killed). I'll continue to monitor and see if I can get a good capture.

 Is it the allocs or heap part you need?
 http://localhost:2019/debug/pprof/allocs?debug=1
 http://localhost:2019/debug/pprof/heap?debug=1

<img width="1440" height="341" alt="Image" src="https://github.com/user-attachments/assets/4657c4d3-3c42-4849-8a4a-d4a2a603a85f" />

`go tool pprof "http://127.0.0.1:2019/debug/pprof/profile?seconds=60` gave this SVG:

<img width="300" height="150" alt="Image" src="https://github.com/user-attachments/assets/72a1835c-acef-4712-8b64-493ee518649d" />

Not sure if it is relevant, but when `SecRuleEngine On`, the following line repeats constantly:
```
{"level":"info","ts":1778880199.882553,"msg":"http: superfluous response.WriteHeader call"}
{"level":"info","ts":1778880200.038215,"msg":"http: superfluous response.WriteHeader call"}
{"level":"info","ts":1778880200.467593,"msg":"http: superfluous response.WriteHeader call"}
{"level":"info","ts":1778880200.7920067,"msg":"http: superfluous response.WriteHeader call"}
{"level":"info","ts":1778880200.9174123,"msg":"http: superfluous response.WriteHeader call"}
```

**Comment by Nixxeee** [comment]
Hello,

I am experiencing the same issue with my Nextcloud vhost.

A few minutes after bringing it online, CPU and RAM usage spikes completely. In my case, this seems to be related to photo synchronization from the iOS app. As soon as the sync starts, the CPU/RAM spike begins.

However, I have already uncommented this block in the `nextcloud-rule-exclusions-config.conf` file:
```
SecAction \
     "id:9508011,\
     phase:1,\
     pass,\
     nolog,\
     ver:'nextcloud-rule-exclusions-plugin/1.6.0',\
     setvar:'tx.nextcloud-rule-exclusions-plugin_ios_workaround_enabled=1'"
```

**Comment by jptosso** [comment]
We tracked this down. The memory blow-up reported here (most recently on Nextcloud sync) is **not** caused by coraza-caddy itself, nor by the per-vhost rule duplication that `memoize` already fixed. It is a request-body handling issue in Coraza, filed upstream as **corazawaf/coraza#1628**.

### What actually happens

Nextcloud clients upload files via WebDAV chunked `PUT` with `Content-Type: application/octet-stream`. With CRS loaded, rule `901340` sets `ctl:forceRequestBodyVariable=On`, and Coraza honors it by forcing the **URLENCODED** body processor for the unknown content type. That copies the entire multi-MB chunk into `REQUEST_BODY` and `ParseQuery`s it into `ARGS_POST`, after which CRS runs its transformation chains (`t:lowercase`, `t:replaceComments`, `t:urlDecodeUni`, …) over the whole binary body. Each transformation allocates a fresh full-size copy, so the cost is roughly `O(body_size × rules)`.

### Performance, measured

Repro: Caddy + coraza-caddy + standard CRS, `SecRuleEngine On`, 12 concurrent ~12 MB octet-stream chunk uploads.

- Process RSS: **~56 MB → ~1.9 GB** (live heap via `:2019/debug/pprof/heap`, GC-independent).
- Per-chunk latency: **~26 s** for a 12 MB body.
- `pprof` heap: **~94%** in `strings.(*Builder)` + `internal/transformations.doReplaceComments` (i.e. the transformation copies), plus `MakeNoZero` from `io.Copy(&strings.Builder, body)` materializing the whole buffered body.

With many concurrent, long-in-flight uploads (exactly the photo/video sync pattern) these short-lived per-transaction copies pile up into the multi-GB usage people are seeing, and Go retains the freed heap as RSS, which is why it "never comes back down".

### Why ModSecurity doesn't have this

This same CRS config is safe under ModSecurity v3 because:

- `ctl:forceRequestBodyVariable` is a **no-op** there, and ModSecurity never url-decodes an octet-stream body into `ARGS` — so the binary body is not parsed into a transformable variable.
- `SecRequestBodyNoFilesLimit` (recommended 128 KB) is **enforced**, capping inspected non-file body. In Coraza this directive is currently parsed but not enforced (corazawaf/coraza#896), and the default `SecRequestBodyLimit` is 128 MB vs ModSecurity's recommended 13 MB.

(Multipart uploads are fine on both engines — file content goes to a temp file / FILES metadata and is never transformed. This is specific to the raw octet-stream WebDAV path.)

### Mitigation today (until the Coraza fix lands)

Keep the WAF active everywhere except request-body inspection on the upload paths, e.g.:

```
SecRule REQUEST_URI "@beginsWith /remote.php/dav" \
    "id:90000,phase:1,nolog,pass,ctl:requestBodyAccess=Off"
```

Place it **after** the CRS includes. In our repro this brought a 12-concurrent-upload run from ~1.9 GB down to ~130 MB flat, with the WAF still blocking attacks on every other path. `SecRequestBodyAccess Off` scoped to the upload vhost is equivalent. (Note: `SecRuleEngine Off` placed *before* `@coraza.conf-recommended` does **not** take effect, because the recommended conf sets `SecRuleEngine DetectionOnly` and directives are last-wins.)

Proper fixes are tracked in corazawaf/coraza#1628 (enforce `SecRequestBodyNoFilesLimit`, default the forced processor to `RAW`, and/or bound transformations by size while still letting libinjection operators inspect the raw body).
